### ISSUE TITLE
VueWrapper made by find() or findAll() can not update parent component

### ISSUE THREAD

**Issue by 38elements** [vuejs/vue-test-utils#284]
component.vue

```vue
<template>
  <div />
</template>

<script>
  export default {
    name: 'component'
  }
</script>
```

component-with-child.vue

```vue
<template>
  <div @click="change">
    <span>
      {{ foo }}
      <child-component ref="child"/>
    </span>
  </div>
</template>

<script>
  import ChildComponent from './component.vue'

  export default{
    name: 'component-with-child',
    data () {
      return {
        foo: 'foo'
      }
    },
    components: {
      ChildComponent
    },
    methods: {
      change () {
        this.foo = 'FOO'
      }
    }
  }
</script>
```

```js
import { mount } from '~vue-test-utils'
import ComponentWithChild from '~resources/components/component-with-child.vue'
import Component from '~resources/components/component.vue'


describe('trigger', () => {
  it('updates all components', () => {
    const wrapper = mount(ComponentWithChild)
    wrapper.find(Component).trigger('click')
    // wrapper.vm.foo is 'FOO'
    expect(wrapper.vm.foo).to.be.equal('FOO')
   // wrapper.text() is  'foo'
    expect(wrapper.text()).to.be.equal('FOO') 
  })
})
```

I think it is a bug.
find() and findAll() return `VueWrapper` object when they are passed a component argument or a ref object.
So, the VueWrapper object has a child component.
VueWrapper.update() can not update the parent component.

**Comment by LinusBorg** [comment]
I think it's just requires a `$nextTick`, so it's not that it "can'T update the parent", it's just that in this case, vue-test-utils doesn't make the update of the DOM synchonous.

**Comment by 38elements** [comment]
IMHO, this may be solved.
The VueWrapper having first argument component of mount()  (root wrapper) has the update method.
This update method is shared with all VueWrapper object.
The child VueWrapper objects do not have the update method for self.

**Comment by LinusBorg** [comment]
That may be true (kind of), but the point is this:

* The click event is triggered in the child.
* For the parent to synchonously update, the child wrapper would have to call the `update method` of the parent wrapper
* While that may technically be possible, this would mean that:
* the parent's parent would also forced to be updated.... and so on.
* ...which means for each of those parents, the `beforeUpdate` and `updated()` hooks would be called, even though that would not happen outside of tests

So in short: a child forcing a synchronous update of a parent would risk to change behaviour in an unintended way, risking false test results or side effects.

**Comment by 38elements** [comment]
Thank you for replay.
There is a possibility of causing problems.
I agree.

**Comment by 38elements** [comment]
In manually, it is resolved as below.

```js
import { mount } from '~vue-test-utils'
import ComponentWithChild from '~resources/components/component-with-child.vue'
import Component from '~resources/components/component.vue'

describe('trigger', () => {
  it('updates all components', () => {
    const wrapper = mount(ComponentWithChild)
    wrapper.find(Component).trigger('click')
    wrapper.update()
    expect(wrapper.vm.foo).to.be.equal('FOO')
    // wrapper.text() is  'FOO'
    expect(wrapper.text()).to.be.equal('FOO')
  })
})
```

**Comment by begueradj** [comment]
What is the difference between `Wrapper` and `VueWrapper` ?

**Comment by dobromir-hristov** [comment]
VueWrapper just wraps a Vue component instance into a Wrapper instance. If its just a normal element (input), it uses just Wrapper.
